### PR TITLE
I217 use a trigger function to assign admins instead of manually doing it

### LINKED ISSUE #217: Use a trigger function to assign admins instead of manually doing it
**Is your feature request related to a problem? Please describe.**
Client has requested to use Firebase Authentication Trigger Functions to assign admins instead of manually assigning users the admin roles.

**Describe the solution you'd like**
Create an auth trigger that will trigger on every user creation using the provided `onCreate` in the firebase functions package.
- When the user is created, use regex to get the domain name of the user's email address.
- If the user domain === whitelisted domain, then set an admin custom token for the user.
- Refer to the `setAdmin` function in the Next API folder.
- If not, do nothing.

The code below is from https://firebase.google.com/docs/auth/admin/custom-claims#logic
```ts
const functions = require('firebase-functions');
const { initializeApp } = require('firebase-admin/app');
const { getAuth } = require('firebase-admin/auth');
const { getDatabase } = require('firebase-admin/database');

initializeApp();

// On sign up.
exports.processSignUp = functions.auth.user().onCreate(async (user) => {
  // Check if user meets role criteria.
  if (
    user.email &&
    user.email.endsWith('@admin.example.com') &&
    user.emailVerified
  ) {
    const customClaims = {
      admin: true,
      accessLevel: 9
    };

    try {
      // Set custom user claims on this newly created user.
      await getAuth().setCustomUserClaims(user.uid, customClaims);

      // Update real-time database to notify client to force refresh.
      const metadataRef = getDatabase().ref('metadata/' + user.uid);

      // Set the refresh time to the current UTC timestamp.
      // This will be captured on the client to force a token refresh.
      await  metadataRef.set({refreshTime: new Date().getTime()});
    } catch (error) {
      console.log(error);
    }
  }
});
```

The client has provided his code below
```ts
/**

 * Splits concatenated config strings into an array of patterns for matching

 */

function getDelimitedConfigPattern(concatenated: string): RegExp {

  return new RegExp(

    concatenated

      ?.split(/[\s,;]+/)

      // Convert '.' delimiters, and '*' wildcard to regex representing valid sub-domain characters

      .map(

        value =>

          '^' + value.trim().replace(/\*\./g, '(*.)?').replace(/\./g, '\\.').replace(/\*/g, '[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?') + '$',

      )

      .filter(value => value?.length)

      .join('|') || '[a-z0-9][a-z0-9-.]*[a-z0-9]',

    'i', // case-insensitive, non-global match

  );

}


/**

 * Called during sign-up and sign-in to stop the auth flow

 * when an incoming user's email domain is not whitelisted.

 */

function validateEmailDomain(email: string | undefined, method: string): void {

  const emailParts = email?.trim().split('@');

  if (emailParts?.length !== 2) {

    throw new functions.auth.HttpsError('permission-denied', 'Sign in email address is invalid', { email });

  }


  // Test against delimited whitelist domains

  const allowedDomains = getDelimitedConfigPattern(whitelist.value());

  const emailDomain = emailParts[1];

  if (!allowedDomains.test(emailDomain) && !allowedDomains.test(email as string)) {

    functions.logger.warn(

      { allowedDomains },

      `${email} ${method.toLowerCase()} blocked - see AzDO Library variable firebase.auth.blocking.domainWhitelist`,

    );

    throw new functions.auth.HttpsError('permission-denied', `${method} is not permitted for ${emailDomain}`, { email });

  }

}
```

### PR DESCRIPTION
**Pull request by dct0**
## Change Summary
**[Briefly summarise the changes that you made. Just high-level stuff]**
Automatically make *@poops.org users admin on signup

### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [X] The pull request title has an issue number
- [X] The change works by "Smoke testing" or quick testing
- [ ] The change has tests
- [X] The change has documentation

## Other Information
**[Is there anything in particular in the review that I should be aware of?]**
I still don't really know the extent of `getDelimitedConfigPattern()` capabilities but hey it works.
WIP will be removed when #200 is merged.

# Related Issue

- Resolve #217
-

### CODE REVIEW

File: firebase/functions/src/trigger/addAdmin.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    throw new functions.auth.HttpsError(
+      'permission-denied',
+      `${method} is not permitted for ${emailDomain}`,
+      { email }
+    )
+  }
+
+  return true
+}
```
Comment (NicholasChoong) [inline firebase/functions/src/trigger/addAdmin.ts:105]: The 2 functions above can be replaced with `user.email.endsWith(<whitelisted_domain>)`.
```ts
exports.processSignUp = functions.auth.user().onCreate(async (user) => {
  // Check if user meets role criteria.
  if (
    user.email &&
    user.email.endsWith(WHITELISTED_DOMAINS) &&
    user.emailVerified
  ) {
   ...
  }
});
``1

**Review by NicholasChoong — CHANGES_REQUESTED** [review]
nice work

File: firebase/functions/src/trigger/addAdmin.ts (hunk 1 above)
Comment (justinliangg) [reply firebase/functions/src/trigger/addAdmin.ts]: Have used a for loop to allow for multiple whitelisted domains

**Review by NicholasChoong — APPROVED** [review]
Nice work

**Pull request merged**